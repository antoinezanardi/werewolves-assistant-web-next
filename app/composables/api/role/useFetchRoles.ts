import type { Role } from "~/composables/api/role/types/role.class";
import { useWerewolvesAssistantApi } from "~/composables/api/useWerewolvesAssistantApi";

type UseFetchRoles = {
  fetchRoles: () => Promise<ReturnType<typeof $fetch<Role[]>> | null>;
};

function useFetchRoles(): UseFetchRoles {
  const { fetchWerewolvesAssistantApi } = useWerewolvesAssistantApi();

  async function fetchRoles(): Promise<ReturnType<typeof $fetch<Role[]>> | null> {
    try {
      return await fetchWerewolvesAssistantApi<Role[]>(`/roles`, { method: "GET" });
    } catch {
      return null;
    }
  }
  return { fetchRoles };
}

export { useFetchRoles };