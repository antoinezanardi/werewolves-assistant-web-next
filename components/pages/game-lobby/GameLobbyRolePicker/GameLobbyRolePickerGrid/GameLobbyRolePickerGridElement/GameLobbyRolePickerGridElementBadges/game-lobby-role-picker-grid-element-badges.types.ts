import type { Role } from "~/composables/api/role/types/role.class";

type GameLobbyRolePickerGridElementBadgesProps = {
  role?: Role;
};

export type { GameLobbyRolePickerGridElementBadgesProps };