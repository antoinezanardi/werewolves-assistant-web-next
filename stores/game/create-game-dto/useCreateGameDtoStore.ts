import { defineStore } from "pinia";
import { ref } from "vue";

import { CreateGamePlayerDto } from "~/composables/api/game/dto/create-game/create-game-player/create-game-player.dto";
import { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import { StoreIds } from "~/stores/enums/store.enum";

const useCreateGameDtoStore = defineStore(StoreIds.CREATE_GAME_DTO, () => {
  const createGameDto = ref<CreateGameDto>(CreateGameDto.create({ players: [] }));

  function setCreateGameDto(createGameDtoValue: CreateGameDto): void {
    createGameDto.value = CreateGameDto.create(createGameDtoValue);
  }

  function resetCreateGameDto(): void {
    createGameDto.value = CreateGameDto.create({ players: [] });
  }

  function addPlayerToCreateGameDto(player: CreateGamePlayerDto): void {
    createGameDto.value.players.push(CreateGamePlayerDto.create(player));
  }

  function updatePlayerInCreateGameDto(player: CreateGamePlayerDto): void {
    const playerIndex = createGameDto.value.players.findIndex(({ name }) => name === player.name);
    if (playerIndex !== -1) {
      createGameDto.value.players.splice(playerIndex, 1, CreateGamePlayerDto.create(player));
    }
  }

  function setPlayersToCreateGameDto(players: CreateGamePlayerDto[]): void {
    createGameDto.value.players = players.map(player => CreateGamePlayerDto.create(player));
  }

  function removePlayerFromCreateGameDto(playerName: string): void {
    const playerIndex = createGameDto.value.players.findIndex(player => player.name === playerName);
    if (playerIndex !== -1) {
      createGameDto.value.players = createGameDto.value.players.toSpliced(playerIndex, 1);
    }
  }
  return {
    createGameDto,
    setCreateGameDto,
    resetCreateGameDto,
    addPlayerToCreateGameDto,
    updatePlayerInCreateGameDto,
    setPlayersToCreateGameDto,
    removePlayerFromCreateGameDto,
  };
});

export { useCreateGameDtoStore };