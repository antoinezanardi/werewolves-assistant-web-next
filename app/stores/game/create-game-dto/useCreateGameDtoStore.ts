import { defineStore } from "pinia";
import { ref } from "vue";
import { DEFAULT_GAME_OPTIONS } from "~/composables/api/game/constants/game-options/game-options.constants";
import type { CreateGameAdditionalCardDto } from "~/composables/api/game/dto/create-game/create-game-additional-card/create-game-additional-card.dto";

import { CreateGamePlayerDto } from "~/composables/api/game/dto/create-game/create-game-player/create-game-player.dto";
import { CreateGameDto } from "~/composables/api/game/dto/create-game/create-game.dto";
import type { GameAdditionalCardRecipientRoleName } from "~/composables/api/game/types/game-additional-card/game-additional-card.types";
import { GameOptions } from "~/composables/api/game/types/game-options/game-options.class";
import { ADDITIONAL_CARDS_DEPENDANT_ROLES } from "~/composables/api/role/constants/role.constants";
import type { RoleName } from "~/composables/api/role/types/role.types";
import { StoreIds } from "~/stores/enums/store.enum";
import { useRolesStore } from "~/stores/role/useRolesStore";
import { useLocalStorage } from "@vueuse/core";
import { LocalStorageKeys } from "~/utils/enums/local-storage.enums";

const useCreateGameDtoStore = defineStore(StoreIds.CREATE_GAME_DTO, () => {
  const rolesStore = useRolesStore();
  const { getRoleWithNameInRoles } = rolesStore;

  const createGameOptionsDtoFromLocalStorage = useLocalStorage(LocalStorageKeys.GAME_OPTIONS, DEFAULT_GAME_OPTIONS, { mergeDefaults: true });

  const defaultCreateGameDto = CreateGameDto.create({
    players: [],
    options: DEFAULT_GAME_OPTIONS,
  });

  const createGameDto = ref<CreateGameDto>(CreateGameDto.create({
    ...defaultCreateGameDto,
    options: createGameOptionsDtoFromLocalStorage.value,
  }));

  const createGameOptionsDto = computed<GameOptions>(() => createGameDto.value.options);

  const doesCreateGameDtoContainPositionDependantRoles = computed<boolean>(() => {
    const positionDependantRoles: RoleName[] = ["rusty-sword-knight", "bear-tamer", "fox"];
    const playersWithPositionDependantRoles = getPlayersWithAnyRoleNameInCreateGameDto(positionDependantRoles);

    return playersWithPositionDependantRoles.length > 0;
  });

  const doesCreateGameDtoContainAdditionalCardsDependantRoles = computed<boolean>(() => {
    const playersWithAdditionalCardsDependantRoles = getPlayersWithAnyRoleNameInCreateGameDto([...ADDITIONAL_CARDS_DEPENDANT_ROLES]);

    return playersWithAdditionalCardsDependantRoles.length > 0;
  });

  function setCreateGameDto(createGameDtoValue: CreateGameDto): void {
    createGameDto.value = CreateGameDto.create(createGameDtoValue);
    saveCreateGameOptionsDtoToLocalStorage();
  }

  function resetCreateGameDto(doesRetrieveLocalStorageValues = true): void {
    if (!doesRetrieveLocalStorageValues) {
      createGameDto.value = CreateGameDto.create(defaultCreateGameDto);
      saveCreateGameOptionsDtoToLocalStorage();

      return;
    }
    createGameDto.value = CreateGameDto.create({
      ...defaultCreateGameDto,
      options: createGameOptionsDtoFromLocalStorage.value,
    });
  }

  function resetCreateGameOptionsDto(): void {
    createGameDto.value.options = GameOptions.create(DEFAULT_GAME_OPTIONS);
    saveCreateGameOptionsDtoToLocalStorage();
  }

  function saveCreateGameOptionsDtoToLocalStorage(): void {
    createGameOptionsDtoFromLocalStorage.value = createGameDto.value.options;
  }

  function removeObsoleteAdditionalCardsFromCreateGameDto(): void {
    if (!createGameDto.value.additionalCards) {
      return;
    }
    const rolesInGame = new Set(createGameDto.value.players.map(player => player.role.name));
    createGameDto.value.additionalCards = createGameDto.value.additionalCards.filter(({ recipient }) => rolesInGame.has(recipient));
  }

  function addPlayerToCreateGameDto(player: CreateGamePlayerDto): void {
    createGameDto.value.players.push(CreateGamePlayerDto.create(player));
  }

  function updatePlayerInCreateGameDto(player: CreateGamePlayerDto): void {
    const playerIndex = createGameDto.value.players.findIndex(({ name }) => name === player.name);
    if (playerIndex !== -1) {
      createGameDto.value.players.splice(playerIndex, 1, CreateGamePlayerDto.create(player));
    }
  }

  function setPlayersToCreateGameDto(players: CreateGamePlayerDto[]): void {
    createGameDto.value.players = players.map(player => CreateGamePlayerDto.create(player));
  }

  function removePlayerFromCreateGameDto(playerName: string): void {
    const playerIndex = createGameDto.value.players.findIndex(player => player.name === playerName);
    if (playerIndex !== -1) {
      createGameDto.value.players = createGameDto.value.players.toSpliced(playerIndex, 1);
    }
  }

  function isRoleInCreateGameDto(roleName: RoleName): boolean {
    return createGameDto.value.players.some(player => player.role.name === roleName);
  }

  function getPlayersWithRoleNameInCreateGameDto(roleName: RoleName): CreateGamePlayerDto[] {
    return createGameDto.value.players.filter(player => player.role.name === roleName);
  }

  function getPlayersWithAnyRoleNameInCreateGameDto(roleNames: RoleName[]): CreateGamePlayerDto[] {
    return createGameDto.value.players.filter(player => player.role.name && roleNames.includes(player.role.name));
  }

  function isRoleMinReachedInCreateGameDto(roleName: RoleName): boolean {
    const playersWithRoleName = getPlayersWithRoleNameInCreateGameDto(roleName);
    const role = getRoleWithNameInRoles(roleName);

    return !!role && (role.minInGame === undefined || playersWithRoleName.length >= role.minInGame);
  }

  function isRoleMaxReachedInCreateGameDto(roleName: RoleName): boolean {
    const playersWithRoleName = getPlayersWithRoleNameInCreateGameDto(roleName);
    const role = getRoleWithNameInRoles(roleName);

    return !!role && playersWithRoleName.length >= role.maxInGame;
  }

  function getRoleLeftCountToReachMinInCreateGameDto(roleName: RoleName): number {
    const playersWithRoleName = getPlayersWithRoleNameInCreateGameDto(roleName);
    const role = getRoleWithNameInRoles(roleName);
    if (role?.minInGame === undefined) {
      return 0;
    }
    const leftCount = role.minInGame - playersWithRoleName.length;

    return leftCount <= -1 ? 0 : leftCount;
  }

  function setAdditionalCardsForRecipientInCreateGameDto(additionalCards: CreateGameAdditionalCardDto[], recipient: GameAdditionalCardRecipientRoleName): void {
    if (!createGameDto.value.additionalCards) {
      createGameDto.value.additionalCards = additionalCards;
    }

    createGameDto.value.additionalCards = createGameDto.value.additionalCards.filter(card => card.recipient !== recipient);
    createGameDto.value.additionalCards.push(...additionalCards);
  }

  function getAdditionalCardsForRecipientInCreateGameDto(recipient: GameAdditionalCardRecipientRoleName): CreateGameAdditionalCardDto[] {
    return createGameDto.value.additionalCards?.filter(card => card.recipient === recipient) ?? [];
  }

  function getAdditionalCardsWithRoleNameInCreateGameDto(roleName: RoleName): CreateGameAdditionalCardDto[] {
    return createGameDto.value.additionalCards?.filter(card => card.roleName === roleName) ?? [];
  }
  return {
    createGameOptionsDtoFromLocalStorage,
    createGameDto,
    createGameOptionsDto,
    doesCreateGameDtoContainPositionDependantRoles,
    doesCreateGameDtoContainAdditionalCardsDependantRoles,
    setCreateGameDto,
    resetCreateGameDto,
    resetCreateGameOptionsDto,
    saveCreateGameOptionsDtoToLocalStorage,
    removeObsoleteAdditionalCardsFromCreateGameDto,
    addPlayerToCreateGameDto,
    updatePlayerInCreateGameDto,
    setPlayersToCreateGameDto,
    removePlayerFromCreateGameDto,
    isRoleInCreateGameDto,
    getPlayersWithRoleNameInCreateGameDto,
    getPlayersWithAnyRoleNameInCreateGameDto,
    isRoleMinReachedInCreateGameDto,
    isRoleMaxReachedInCreateGameDto,
    getRoleLeftCountToReachMinInCreateGameDto,
    setAdditionalCardsForRecipientInCreateGameDto,
    getAdditionalCardsForRecipientInCreateGameDto,
    getAdditionalCardsWithRoleNameInCreateGameDto,
  };
});

export { useCreateGameDtoStore };