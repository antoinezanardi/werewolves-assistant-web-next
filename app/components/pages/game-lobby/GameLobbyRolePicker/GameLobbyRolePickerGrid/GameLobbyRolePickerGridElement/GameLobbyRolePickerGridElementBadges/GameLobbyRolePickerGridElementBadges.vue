<template>
  <div
    id="game-lobby-role-picker-grid-element-badges"
  >
    <PrimeVueBadge
      v-if="countInGameAdditionalCardsCreateGameDto"
      id="role-count-in-additional-cards-badge"
      v-p-tooltip="$t('components.GameLobbyRolePickerGridElementBadges.totalInAdditionalCards')"
      class="-right-4 absolute top-1"
      data-testid="game-lobby-role-picker-role-count-in-additional-cards-badge"
      severity="warning"
    >
      <i class="fa fa-clover me-2"/>

      <span>
        {{ countInGameAdditionalCardsCreateGameDto }}
      </span>
    </PrimeVueBadge>

    <PrimeVueBadge
      v-if="countInCreateGameDto"
      id="role-count-in-game-badge"
      v-p-tooltip="$t('components.GameLobbyRolePickerGridElementBadges.totalInParty')"
      class="-right-4 absolute top-1"
      data-testid="game-lobby-role-picker-role-count-in-game-badge"
    >
      <i class="fa fa-chess-pawn me-2"/>

      <span>
        {{ countInCreateGameDto }}
      </span>
    </PrimeVueBadge>

    <PrimeVueBadge
      v-if="countInCreateGameDto && leftCountToReachMinInCreateGameDto"
      id="role-min-count-in-game-badge"
      v-p-tooltip="$t('components.GameLobbyRolePickerGridElementBadges.minInGameNotReached', { 'count': leftCountToReachMinInCreateGameDto })"
      class="-right-4 absolute top-8"
      data-testid="game-lobby-role-picker-role-min-count-in-game-badge"
      severity="danger"
    >
      <i class="fa fa-exclamation-circle"/>
    </PrimeVueBadge>
  </div>
</template>

<script setup lang="ts">
import type { GameLobbyRolePickerGridElementBadgesProps } from "~/components/pages/game-lobby/GameLobbyRolePicker/GameLobbyRolePickerGrid/GameLobbyRolePickerGridElement/GameLobbyRolePickerGridElementBadges/game-lobby-role-picker-grid-element-badges.types";
import { useCreateGameDtoStore } from "~/stores/game/create-game-dto/useCreateGameDtoStore";

const props = defineProps<GameLobbyRolePickerGridElementBadgesProps>();

const createGameDtoStore = useCreateGameDtoStore();
const {
  getPlayersWithRoleNameInCreateGameDto,
  getRoleLeftCountToReachMinInCreateGameDto,
  getAdditionalCardsWithRoleNameInCreateGameDto,
} = createGameDtoStore;

const countInCreateGameDto = computed<number>(() => {
  if (!props.role) {
    return 0;
  }
  return getPlayersWithRoleNameInCreateGameDto(props.role.name).length;
});

const leftCountToReachMinInCreateGameDto = computed<number>(() => {
  if (!props.role) {
    return 0;
  }
  return getRoleLeftCountToReachMinInCreateGameDto(props.role.name);
});

const countInGameAdditionalCardsCreateGameDto = computed<number>(() => {
  if (!props.role) {
    return 0;
  }
  return getAdditionalCardsWithRoleNameInCreateGameDto(props.role.name).length;
});
</script>