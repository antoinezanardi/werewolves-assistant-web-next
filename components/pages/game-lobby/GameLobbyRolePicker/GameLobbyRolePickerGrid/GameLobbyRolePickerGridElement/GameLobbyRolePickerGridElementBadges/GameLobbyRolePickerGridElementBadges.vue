<template>
  <div
    id="game-lobby-role-picker-grid-element-badges"
  >
    <VuePrimeBadge
      v-if="countInCreateGameDto"
      id="role-count-in-game-badge"
      v-p-tooltip="$t('components.GameLobbyRolePickerGridElementBadges.totalInParty')"
      class="-right-4 absolute top-1"
    >
      <i class="fa fa-chess-pawn me-2"/>

      <span>
        {{ countInCreateGameDto }}
      </span>
    </VuePrimeBadge>

    <VuePrimeBadge
      v-if="countInCreateGameDto && leftCountToReachMinInCreateGameDto"
      id="role-min-count-in-game-badge"
      v-p-tooltip="$t('components.GameLobbyRolePickerGridElementBadges.minInGameNotReached', { 'count': leftCountToReachMinInCreateGameDto })"
      class="-right-4 absolute top-8"
      severity="danger"
    >
      <i class="fa fa-exclamation-circle"/>
    </VuePrimeBadge>
  </div>
</template>

<script setup lang="ts">
import type { GameLobbyRolePickerGridElementBadgesProps } from "~/components/pages/game-lobby/GameLobbyRolePicker/GameLobbyRolePickerGrid/GameLobbyRolePickerGridElement/GameLobbyRolePickerGridElementBadges/game-lobby-role-picker-grid-element-badges.types";
import { useCreateGameDtoStore } from "~/stores/game/create-game-dto/useCreateGameDtoStore";

const props = defineProps<GameLobbyRolePickerGridElementBadgesProps>();

const createGameDtoStore = useCreateGameDtoStore();
const {
  getPlayersWithRoleNameInCreateGameDto,
  getRoleLeftCountToReachMinInCreateGameDto,
} = createGameDtoStore;

const countInCreateGameDto = computed<number>(() => {
  if (!props.role) {
    return 0;
  }
  return getPlayersWithRoleNameInCreateGameDto(props.role.name).length;
});

const leftCountToReachMinInCreateGameDto = computed<number>(() => {
  if (!props.role) {
    return 0;
  }
  return getRoleLeftCountToReachMinInCreateGameDto(props.role.name);
});
</script>