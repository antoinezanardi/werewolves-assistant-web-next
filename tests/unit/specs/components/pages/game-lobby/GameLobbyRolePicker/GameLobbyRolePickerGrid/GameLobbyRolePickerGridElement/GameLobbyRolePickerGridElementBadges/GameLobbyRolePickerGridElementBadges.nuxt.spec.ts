import { createTestingPinia } from "@pinia/testing";
import type { mount } from "@vue/test-utils";
import type { ComponentMountingOptions } from "@vue/test-utils/dist/mount";
import type Badge from "primevue/badge";

import { createFakeCreateGameAdditionalCardDto } from "@tests/unit/utils/factories/composables/api/game/dto/create-game/create-game-additional-card/create-game-additional-card.dto.factory";
import type { GameLobbyRolePickerGridElementBadgesProps } from "~/components/pages/game-lobby/GameLobbyRolePicker/GameLobbyRolePickerGrid/GameLobbyRolePickerGridElement/GameLobbyRolePickerGridElementBadges/game-lobby-role-picker-grid-element-badges.types";
import GameLobbyRolePickerGridElementBadges from "~/components/pages/game-lobby/GameLobbyRolePicker/GameLobbyRolePickerGrid/GameLobbyRolePickerGridElement/GameLobbyRolePickerGridElementBadges/GameLobbyRolePickerGridElementBadges.vue";
import { useCreateGameDtoStore } from "~/stores/game/create-game-dto/useCreateGameDtoStore";
import { createFakeCreateGamePlayerDto } from "@tests/unit/utils/factories/composables/api/game/dto/create-game/create-game-player/create-game-player.dto.factory";
import { createFakeRole } from "@tests/unit/utils/factories/composables/api/role/role.factory";
import { pTooltipDirectiveBinder } from "@tests/unit/utils/helpers/directive.helpers";
import { mockPiniaStore } from "@tests/unit/utils/helpers/mock.helpers";
import { mountSuspendedComponent } from "@tests/unit/utils/helpers/mount.helpers";
import type { BoundTooltip } from "@tests/unit/utils/types/directive.types";

describe("Game Lobby Role Picker Grid Element Badges Component", () => {
  let wrapper: ReturnType<typeof mount<typeof GameLobbyRolePickerGridElementBadges>>;
  const defaultProps: GameLobbyRolePickerGridElementBadgesProps = { role: createFakeRole({ name: "seer" }) };

  async function mountGameLobbyRolePickerGridElementBadgesComponent(options: ComponentMountingOptions<typeof GameLobbyRolePickerGridElementBadges>):
  Promise<ReturnType<typeof mount<typeof GameLobbyRolePickerGridElementBadges>>> {
    return mountSuspendedComponent(GameLobbyRolePickerGridElementBadges, {
      props: defaultProps,
      ...options,
    });
  }

  beforeEach(async() => {
    const pinia = createTestingPinia();
    const createGameDtoStore = mockPiniaStore(useCreateGameDtoStore);
    createGameDtoStore.getPlayersWithRoleNameInCreateGameDto.mockReturnValue([createFakeCreateGamePlayerDto()]);
    createGameDtoStore.getRoleLeftCountToReachMinInCreateGameDto.mockReturnValue(1);
    createGameDtoStore.getAdditionalCardsWithRoleNameInCreateGameDto.mockReturnValue([]);
    wrapper = await mountGameLobbyRolePickerGridElementBadgesComponent({
      shallow: false,
      global: { plugins: [pinia] },
    });
  });

  it("should match snapshot when rendered.", () => {
    expect(wrapper).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  });

  describe("Count in Additional Cards Badge", () => {
    it("should not render the count in additional cards badge when there are no additional cards with this role.", () => {
      const countInAdditionalCardsBadge = wrapper.findComponent<typeof Badge>("#role-count-in-additional-cards-badge");

      expect(countInAdditionalCardsBadge.exists()).toBeFalsy();
    });

    it("should assign tooltip to count in additional cards badge when there are additional cards with this role.", async() => {
      const tooltip: BoundTooltip = { value: undefined };
      const directives = { ...pTooltipDirectiveBinder(tooltip, "#role-count-in-additional-cards-badge") };
      const pinia = createTestingPinia();
      const createGameDtoStore = mockPiniaStore(useCreateGameDtoStore);
      createGameDtoStore.getPlayersWithRoleNameInCreateGameDto.mockReturnValue([createFakeCreateGamePlayerDto()]);
      createGameDtoStore.getRoleLeftCountToReachMinInCreateGameDto.mockReturnValue(1);
      createGameDtoStore.getAdditionalCardsWithRoleNameInCreateGameDto.mockReturnValue([createFakeCreateGameAdditionalCardDto()]);
      wrapper = await mountGameLobbyRolePickerGridElementBadgesComponent({
        global: {
          plugins: [pinia],
          directives,
        },
      });

      expect(tooltip.value).toBe("Total in the additional cards");
    });

    it("should not render the count in additional cards badge when role is not defined.", async() => {
      const pinia = createTestingPinia();
      const createGameDtoStore = mockPiniaStore(useCreateGameDtoStore);
      createGameDtoStore.getPlayersWithRoleNameInCreateGameDto.mockReturnValue([createFakeCreateGamePlayerDto()]);
      createGameDtoStore.getRoleLeftCountToReachMinInCreateGameDto.mockReturnValue(1);
      createGameDtoStore.getAdditionalCardsWithRoleNameInCreateGameDto.mockReturnValue([createFakeCreateGameAdditionalCardDto()]);
      wrapper = await mountGameLobbyRolePickerGridElementBadgesComponent({
        global: { plugins: [pinia] },
        props: { role: undefined },
      });
      const countInAdditionalCardsBadge = wrapper.findComponent<typeof Badge>("#role-count-in-additional-cards-badge");

      expect(countInAdditionalCardsBadge.exists()).toBeFalsy();
    });

    it("should set count in game additional cards create game dto text when there are additional cards with this role.", async() => {
      const pinia = createTestingPinia();
      const createGameDtoStore = mockPiniaStore(useCreateGameDtoStore);
      createGameDtoStore.getPlayersWithRoleNameInCreateGameDto.mockReturnValue([createFakeCreateGamePlayerDto()]);
      createGameDtoStore.getRoleLeftCountToReachMinInCreateGameDto.mockReturnValue(1);
      createGameDtoStore.getAdditionalCardsWithRoleNameInCreateGameDto.mockReturnValue([createFakeCreateGameAdditionalCardDto()]);
      wrapper = await mountGameLobbyRolePickerGridElementBadgesComponent({
        shallow: false,
        global: {
          plugins: [pinia],
        },
      });
      const countInAdditionalCardsBadge = wrapper.findComponent<typeof Badge>("#role-count-in-additional-cards-badge");

      expect(countInAdditionalCardsBadge.text()).toBe("1");
    });
  });

  describe("Count in Party Badge", () => {
    it("should render the count in party badge when there are players with this role.", () => {
      const countInPartyBadge = wrapper.findComponent<typeof Badge>("#role-count-in-game-badge");

      expect(countInPartyBadge.exists()).toBeTruthy();
    });

    it("should assign tooltip to count in party badge when there are players with this role.", async() => {
      const tooltip: BoundTooltip = { value: undefined };
      const directives = { ...pTooltipDirectiveBinder(tooltip, "#role-count-in-game-badge") };
      const pinia = createTestingPinia();
      const createGameDtoStore = mockPiniaStore(useCreateGameDtoStore);
      createGameDtoStore.getPlayersWithRoleNameInCreateGameDto.mockReturnValue([createFakeCreateGamePlayerDto()]);
      createGameDtoStore.getRoleLeftCountToReachMinInCreateGameDto.mockReturnValue(1);
      createGameDtoStore.getAdditionalCardsWithRoleNameInCreateGameDto.mockReturnValue([]);
      wrapper = await mountGameLobbyRolePickerGridElementBadgesComponent({
        global: {
          plugins: [pinia],
          directives,
        },
      });

      expect(tooltip.value).toBe("Total in the party");
    });

    it("should not render the count in party badge when there are no players with this role.", async() => {
      const pinia = createTestingPinia();
      const createGameDtoStore = mockPiniaStore(useCreateGameDtoStore);
      createGameDtoStore.getPlayersWithRoleNameInCreateGameDto.mockReturnValue([]);
      createGameDtoStore.getAdditionalCardsWithRoleNameInCreateGameDto.mockReturnValue([]);
      wrapper = await mountGameLobbyRolePickerGridElementBadgesComponent({ global: { plugins: [pinia] } });
      const countInPartyBadge = wrapper.findComponent<typeof Badge>("#role-count-in-game-badge");

      expect(countInPartyBadge.exists()).toBeFalsy();
    });

    it("should not render the count in party badge when role is not defined.", async() => {
      const pinia = createTestingPinia();
      const createGameDtoStore = mockPiniaStore(useCreateGameDtoStore);
      createGameDtoStore.getPlayersWithRoleNameInCreateGameDto.mockReturnValue([createFakeCreateGamePlayerDto()]);
      createGameDtoStore.getRoleLeftCountToReachMinInCreateGameDto.mockReturnValue(1);
      createGameDtoStore.getAdditionalCardsWithRoleNameInCreateGameDto.mockReturnValue([]);
      wrapper = await mountGameLobbyRolePickerGridElementBadgesComponent({
        global: { plugins: [pinia] },
        props: { role: undefined },
      });
      const countInPartyBadge = wrapper.findComponent<typeof Badge>("#role-count-in-game-badge");

      expect(countInPartyBadge.exists()).toBeFalsy();
    });
  });

  describe("Left to Pick Badge", () => {
    it("should render the left to pick badge when there are roles left to pick.", () => {
      const leftToPickBadge = wrapper.findComponent<typeof Badge>("#role-min-count-in-game-badge");

      expect(leftToPickBadge.exists()).toBeTruthy();
    });

    it("should assign tooltip to left to pick badge when there are roles left to pick.", async() => {
      const tooltip: BoundTooltip = { value: undefined };
      const directives = { ...pTooltipDirectiveBinder(tooltip, "#role-min-count-in-game-badge") };
      const pinia = createTestingPinia();
      const createGameDtoStore = mockPiniaStore(useCreateGameDtoStore);
      createGameDtoStore.getPlayersWithRoleNameInCreateGameDto.mockReturnValue([createFakeCreateGamePlayerDto()]);
      createGameDtoStore.getRoleLeftCountToReachMinInCreateGameDto.mockReturnValue(1);
      createGameDtoStore.getAdditionalCardsWithRoleNameInCreateGameDto.mockReturnValue([]);
      wrapper = await mountGameLobbyRolePickerGridElementBadgesComponent({
        global: {
          plugins: [pinia],
          directives,
        },
      });

      expect(tooltip.value).toBe("Minimum in the party not reached (1 left to pick)");
    });

    it("should not render the left to pick badge when there are no roles left to pick.", async() => {
      const pinia = createTestingPinia();
      const createGameDtoStore = mockPiniaStore(useCreateGameDtoStore);
      createGameDtoStore.getPlayersWithRoleNameInCreateGameDto.mockReturnValue([createFakeCreateGamePlayerDto()]);
      createGameDtoStore.getRoleLeftCountToReachMinInCreateGameDto.mockReturnValue(0);
      createGameDtoStore.getAdditionalCardsWithRoleNameInCreateGameDto.mockReturnValue([]);
      wrapper = await mountGameLobbyRolePickerGridElementBadgesComponent({ global: { plugins: [pinia] } });
      const leftToPickBadge = wrapper.findComponent<typeof Badge>("#role-min-count-in-game-badge");

      expect(leftToPickBadge.exists()).toBeFalsy();
    });

    it("should not render the left to pick badge when there are no players with this role.", async() => {
      const pinia = createTestingPinia();
      const createGameDtoStore = mockPiniaStore(useCreateGameDtoStore);
      createGameDtoStore.getPlayersWithRoleNameInCreateGameDto.mockReturnValue([]);
      createGameDtoStore.getRoleLeftCountToReachMinInCreateGameDto.mockReturnValue(1);
      createGameDtoStore.getAdditionalCardsWithRoleNameInCreateGameDto.mockReturnValue([]);
      wrapper = await mountGameLobbyRolePickerGridElementBadgesComponent({ global: { plugins: [pinia] } });
      const leftToPickBadge = wrapper.findComponent<typeof Badge>("#role-min-count-in-game-badge");

      expect(leftToPickBadge.exists()).toBeFalsy();
    });

    it("should not render the left to pick badge when role is not defined.", async() => {
      const pinia = createTestingPinia();
      const createGameDtoStore = mockPiniaStore(useCreateGameDtoStore);
      createGameDtoStore.getPlayersWithRoleNameInCreateGameDto.mockReturnValue([createFakeCreateGamePlayerDto()]);
      createGameDtoStore.getRoleLeftCountToReachMinInCreateGameDto.mockReturnValue(1);
      createGameDtoStore.getAdditionalCardsWithRoleNameInCreateGameDto.mockReturnValue([]);
      wrapper = await mountGameLobbyRolePickerGridElementBadgesComponent({
        global: { plugins: [pinia] },
        props: { role: undefined },
      });
      const leftToPickBadge = wrapper.findComponent<typeof Badge>("#role-min-count-in-game-badge");

      expect(leftToPickBadge.exists()).toBeFalsy();
    });
  });
});