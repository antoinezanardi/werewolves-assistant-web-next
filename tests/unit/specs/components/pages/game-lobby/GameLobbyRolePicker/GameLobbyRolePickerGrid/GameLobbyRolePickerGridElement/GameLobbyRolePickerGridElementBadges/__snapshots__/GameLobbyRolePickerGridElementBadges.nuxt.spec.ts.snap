// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Game Lobby Role Picker Grid Element Badges Component > should match snapshot when rendered. 1`] = `
"<div id="game-lobby-role-picker-grid-element-badges">
  <!--v-if--><span class="p-badge p-component -right-4 absolute top-1" id="role-count-in-game-badge" data-testid="game-lobby-role-picker-role-count-in-game-badge" data-pc-name="badge" pc1="" data-pc-section="root" data-pd-tooltip="true"><svg class="svg-inline--fa fa-chess-pawn me-1" aria-hidden="true" focusable="false" data-prefix="fas" data-icon="chess-pawn" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 320 512"><path class="" fill="currentColor" d="M215.5 224c29.2-18.4 48.5-50.9 48.5-88c0-57.4-46.6-104-104-104S56 78.6 56 136c0 37.1 19.4 69.6 48.5 88L96 224c-17.7 0-32 14.3-32 32c0 16.5 12.5 30 28.5 31.8L80 400l160 0L227.5 287.8c16-1.8 28.5-15.3 28.5-31.8c0-17.7-14.3-32-32-32l-8.5 0zM22.6 473.4c-4.2 4.2-6.6 10-6.6 16C16 501.9 26.1 512 38.6 512l242.7 0c12.5 0 22.6-10.1 22.6-22.6c0-6-2.4-11.8-6.6-16L256 432 64 432 22.6 473.4z"></path></svg><span>1</span></span><span class="p-badge p-component p-badge-danger -right-4 absolute top-8" id="role-min-count-in-game-badge" data-testid="game-lobby-role-picker-role-min-count-in-game-badge" data-pc-name="badge" pc2="" data-pc-section="root" data-pd-tooltip="true"><svg class="svg-inline--fa fa-exclamation-circle" aria-hidden="true" focusable="false" data-prefix="fas" data-icon="exclamation-circle" role="img" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><path class="" fill="currentColor" d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zm0-384c13.3 0 24 10.7 24 24l0 112c0 13.3-10.7 24-24 24s-24-10.7-24-24l0-112c0-13.3 10.7-24 24-24zM224 352a32 32 0 1 1 64 0 32 32 0 1 1 -64 0z"></path></svg></span>
</div>"
`;
